### PR TITLE
feat(ui): add port mapping table

### LINKED ISSUE #1682: feat(ui): add port mapping table based on selected config port

### PR DESCRIPTION
**Pull request by ReenigneArcher**
## Description

Add a port configuration table to the UI.

Also validates the config port is is in the range of 1029 to 65514, and that all ports are in the range of 1024 to 65535.

Modifies alerts to include a font awesome icon.

### Screenshot

Default value:
image

Custom value:
image

Invalid values:
image
image

### Issues Fixed or Closed

- Resolves #1682 

## Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update (updates to dependencies)
- [ ] Documentation update (changes to documentation)
- [ ] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

## Branch Updates
LizardByte requires that branches be up-to-date before merging. This means that after any PR is merged, this branch
must be updated before it can be merged. You must also
[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I want maintainers to keep my branch updated

### CODE REVIEW

**Comment by cgutman** [comment]
Can you add a description or some emphasis on the HTTP port entry that indicates that's the one that should be provided to Moonlight when adding a PC manually?

**Comment by yungleballz** [comment]
I would like to suggest to include an info that the Web UI Port does not have to (and should not) be forwarded.

**Comment by ReenigneArcher** [comment]
@cgutman yes, good idea.

@yungleballz that will depend on user preferences. Many users forward it so they can modify settings away from home, or pair a new client.

**Comment by ReenigneArcher** [comment]
Idea: Add port checker -> https://www.npmjs.com/package/is-port-reachable

**Comment by TheElixZammuto** [comment]
> @yungleballz that will depend on user preferences. Many users forward it so they can modify settings away from home, or pair a new client.

I agree with yungle here, we need a big red scary warning to avoid people forwarding the UI without knowing what they do

> Idea: Add port checker -> https://www.npmjs.com/package/is-port-reachable
> 
> This probably needs to be in a separate PR after webpack is implemented.

Due to NAT (and hairpin), having Sunshine itself make the test is not representitive, the Internet Streaming Tool uses a relay server IIRC

File: src/main.cpp [hunk 1]
```diff
@@ -836,7 +836,13 @@ write_file(const char *path, const std::string_view &contents) {
  */
 std::uint16_t
 map_port(int port) {
-  // TODO: Ensure port is in the range of 21-65535
+  // Ensure port is in the range of 1024-65535
+  if (port < 1024 || port > 65535) {
+    BOOST_LOG(warning) << "Port out of range: "sv << port;
```
Comment (TheElixZammuto) [inline src/main.cpp:841]: While ports < 1024 are root-only, these are still valid ports. I think in this specific case making the warning is enough

File: src_assets/common/assets/web/config.html [hunk 2]
```diff
@@ -625,14 +625,74 @@ <h1 class="my-4">Configuration</h1>
...
+              <td>UDP</td>
+              <td>{{+effectivePort + 11}}</td>
+            </tr>
+            <tr>
+              <td>Mic (unused)</td>
+              <td>UDP</td>
+              <td>{{+effectivePort + 13}}</td>
+            </tr>
+          </tbody>
+        </table>
```
Comment (TheElixZammuto) [inline src_assets/common/assets/web/config.html:695]: I would avoid to specify which port are what for. People would wrongly assume that they can choose which port might enable, while it's not the case

**Comment by ReenigneArcher** [comment]
> Due to NAT (and hairpin), having Sunshine itself make the test is not representitive, the Internet Streaming Tool uses a relay server IIRC

I wanted to add two columns.

- Show if port is available on the localhost
- Show if port is available using public ip address (maybe we can use this site somehow? https://canyouseeme.org/)

File: src_assets/common/assets/web/config.html [hunk 3]
```diff
@@ -625,14 +625,74 @@ <h1 class="my-4">Configuration</h1>
...
           v-model="config.port"
         />
         <div class="form-text">Set the family of ports used by Sunshine</div>
+        <!-- Add warning if any port is less than 1024 -->
+        <div class="alert alert-danger" v-if="(+effectivePort - 5) < 1024">
+          <b>Warning:</b> Ports below 1024 are not recommended!
+        </div>
+        <!-- Add warning if any port is above 65535 -->
+        <div class="alert alert-danger" v-if="(+effectivePort + 21) > 65535">
+          <b>Danger:</b> Ports above 65535 are not available!
```
Comment (TheElixZammuto) [inline src_assets/common/assets/web/config.html:642]:
```suggestion
          <b>Error:</b> Ports above 65535 are not available!
```

**Review by TheElixZammuto — CHANGES_REQUESTED** [review]

File: src/main.cpp (hunk 1 above)
Comment (ReenigneArcher) [reply src/main.cpp:841]: This condition should actually never happen, as I modified the range for config port 1029-65514 (to account for the difference from the config port (-5, +21)

**Comment by ReenigneArcher** [comment]
@cgutman This will appear if the user specifies a non default port.

image

**Comment by ReenigneArcher** [comment]
Modified UI to:

Default:
image

Non Default:
image

Too Low:
image

Too High:
image

**Comment by ReenigneArcher** [comment]
Added these warnings if web ui is allowed on WAN.

image

image

**Review by TheElixZammuto — APPROVED** [review]

**Pull request merged**